src/box/tuple_compare.cc

```diff
@@ -1610,6 +1647,25 @@ hint_uuid_raw(const char *data)
 	return hint_create(MP_CLASS_UUID, val);
 }
 
+static inline hint_t
+hint_datetime(struct datetime *date)
+{
+	/*
+	 * Use at most HINT_VALUE_SECS_BITS from datetime
+	 * seconds field as a hint value, and at MSB part
+	 * of HINT_VALUE_NSEC_BITS from nanoseconds.
+	 */
+	int64_t secs = date->epoch;
+	int32_t nsec = date->nsec;
+	uint64_t val = secs <= HINT_VALUE_SECS_MIN ? 0 :
```

Comment:
I see a serious problem here. We must not bound high part of hint independently on low part.
Imagine two dates:
1) sec = 0xFFFFFFFFF01, nsec = 2
2) sec = 0xFFFFFFFFF02, nsec = 1
Obviously the second is greater.
But when you apply your hint algo, you'll have something like
1) hint = 0xFFFFFFFFF000002
2) hint = 0xFFFFFFFFF000001
And you'll get the first hint greater that the second, which is wrong.
I would be glad if you add a test for that.

Reply:
Yeah, that's off if comparison by nanoseconds would be providing erroneous results in case of underflow/overflow.

Idea is - to not save nsec part of there is any of underflow/overflow of epoch field occurred, saving 0 to nsec part (making them equal)

Reply:
Here is corrected code for datetime hint calculation, which would be properly underflow/overflow on boundary conditions. (And not set nanoseconds parts if there is overflow)

```diff
@@ -1656,13 +1656,25 @@ hint_datetime(struct datetime *date)
         * of HINT_VALUE_NSEC_BITS from nanoseconds.
         */
        int64_t secs = date->epoch;
-       int32_t nsec = date->nsec;
-       uint64_t val = secs <= HINT_VALUE_SECS_MIN ? 0 :
-                       secs - HINT_VALUE_SECS_MIN;
-       if (val >= HINT_VALUE_SECS_MAX)
+       /*
+        * overflow or underflow
+        */
+       bool overflow = false;
+       uint64_t val;
+       if (secs < HINT_VALUE_SECS_MIN) {
+               overflow = true;
+               val = 0;
+       } else {
+               val = secs - HINT_VALUE_SECS_MIN;
+       }
+       if (val > HINT_VALUE_SECS_MAX) {
+               overflow = true;
                val = HINT_VALUE_SECS_MAX;
+       }
        val <<= HINT_VALUE_NSEC_BITS;
-       val |= (nsec >> HINT_VALUE_NSEC_SHIFT) & HINT_VALUE_NSEC_MAX;
+       if (overflow == false)
+               val |= (date->nsec >> HINT_VALUE_NSEC_SHIFT) &
+                       HINT_VALUE_NSEC_MAX;
        return hint_create(MP_CLASS_DATETIME, val);
 }
```